onnxruntime/core/providers/webnn/builders/impl/flatten_op_builder.cc

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Intel Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/webnn/builders/helper.h"
+#include "core/providers/webnn/builders/model_builder.h"
+#include "core/providers/webnn/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace webnn {
+
+class FlattenOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+};
+
+// Add operator related.
+
+Status FlattenOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                               const Node& node,
+                                               const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  ORT_RETURN_IF_NOT(input_defs.size() < 1, "Flatten has no input tensor");
+  if (!GetShape(*input_defs[0], input_shape, logger)) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "FlattenOpBuilder::AddToModelBuilderImpl, cannot get input shape");
+  }
+  auto rank = input_shape.size();
+  NodeAttrHelper helper(node);
+  uint32_t axis = static_cast<uint32_t>(HandleNegativeAxis(helper.Get("axis", 1), rank));
+
+  emscripten::val inputs = model_builder.GetOperand(input_defs[0]->Name());
+
+  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("flatten", inputs, axis);
```

Comment:
```suggestion
  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("flattenTo2d", inputs, axis);
```

It takes an ND tensor and flattens it to a 2D tensor.

```
interface MLGraphBuilder {
  ...
  [RaisesException] MLOperand flattenTo2d(MLOperand input, unsigned long axis);
  ..
}
```

https://github.com/fdwr/chromium-src-webnn-dml/pull/1/files#diff-e1b2517a6ae8f7c4494c75d17c8650b56e4f8d430f54f5e1f765475f00a5e1f3R231

Thanks Wang.